### DISCUSSION TITLE
How to show an image from disk?

### DISCUSSION THREAD

**Discussion by 200729** [zauberzeug/nicegui#28] [Q&A] [answered]
Can image be loaded from disk (not http)?
ui.image('image.jpg')
I tried but image has not been loaded, I tried also absolute path.

**Comment by falkoschindler** [comment, marked as answer]
Yes, you can load an image from disk - somehow.

NiceGUI provides both frontend and backend of a web application. The frontend is running in the browser - possibly on a different machine than the server. Therefore the frontend code cannot simply refer to a local image on the server's machine.

But you can *serve* the image and then refer to it via URL:

```python
#!/usr/bin/env python3
from nicegui import ui
from starlette import responses

@ui.get('/{filename}')
def get_file(filename: str):
    return responses.FileResponse(filename)

ui.image('image.jpg')

ui.run()
```

Here we define an endpoint `get_file` that returns an arbitrary file with the given `filename`. A more specific or even more general implementation is also possible.

We will have to think about if this should be more built-in in NiceGUI, like `ui.local_image(...)`. On the one hand it's good to serve files explicitly instead of secretly giving remote access to the local filesystem. One the other hand requiring 4 lines of code to insert an image from disk is not very *nice*.

---

**Update:**

You can also use Starlette to mount a static file directory, e.g. under the path "/images":

```python
#!/usr/bin/env python3
from nicegui import ui
from starlette.routing import Mount
from starlette.staticfiles import StaticFiles

ui.add_route(Mount('/images', app=StaticFiles(directory='.')))

ui.image('/images/image.jpg')

ui.run()
```

---

**Yet another update:**

Since version 0.7.28 you can use `ui.add_static_files`:

```python
#!/usr/bin/env python3
from nicegui import ui

ui.add_static_files('/images', '.')

ui.image('/images/image.jpg')

ui.run()
```

**Reply by 200729** [reply]
Mount a static file directory with Starlette is working, thank you!

**Reply by canDry** [reply]
**Correction**:  add_static_files is part of app (not ui)
```python
from nicegui import ui, app
app.add_static_files('/images', '.')

ui.image('/images/image.jpg')

ui.run()
```
Also note: it works for audio as well as images.

**Reply by JustinHowes01** [reply]
This method does not seem to work now. Is there an updated way to do this?

**Reply by rodja** [reply]
I just've run [the slideshow example](https://github.com/zauberzeug/nicegui/blob/main/examples/slideshow/main.py) again without any problems. Maybe you can start a [new discussion](https://github.com/zauberzeug/nicegui/discussions/categories/q-a) providing code and your specific setup...

**Reply by JustinHowes01** [reply]
I see from the slideshow example that you used `app.add_static_files('/images', folder)` instead of the method above. That worked in my case! Thank you for providing the example

**Reply by rodja** [reply]
I've just tried `app.add_static_files('/images', '.')` and it works as expected. All files in the working dir are served. That means the script must be started from the folder where the `image.jpg` is located. To start from anywhere you need to compute the path relative from the main script location as we do in 

https://github.com/zauberzeug/nicegui/blob/b2b9687b9256e128e4b431139d8f25a072583770/examples/slideshow/main.py#L7

**Comment by rodja** [comment]
We just released [v1.2.20](https://github.com/zauberzeug/nicegui/releases/tag/v1.2.20) which also introduces `app.add_media_files` for streaming support and also allows you to pass a local file directly to `ui.video`, `ui.audio`, `ui.image` and `ui.interactive_image`. It will automatically be served correctly.
